packages/odf-plugin-sdk/extensions/external-storage/external-storage.ts

```diff
@@ -0,0 +1,102 @@
+import { K8sResourceKind } from '@odf/shared/types';
+import { K8sModel } from '@openshift-console/dynamic-plugin-sdk/lib/api/common-types';
+import {
+  Extension,
+  ExtensionDeclaration,
+} from '@openshift-console/dynamic-plugin-sdk/lib/types';
+import { Control } from 'react-hook-form';
+
+type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
+
+type Model = PartialBy<K8sModel, 'abbr' | 'label' | 'labelPlural'>;
+
+/**
+ *  Configures a new external storage vendor to the Storage System Creation wizard.
+ */
+export type StorageClassWizardStepProps = {
+  /** Display name of the external storage vendor. */
+  displayName: string;
+  /** The model referring the `apiGroup`,`apiVersion`, `plural` and `kind` of the external storage vendor's CRD. */
+  model: Model;
+  /** A React Functional Component to input the connection details of the external storage vendor. */
+  component: React.FunctionComponent<StorageClassComponentProps<{}>>;
+  /**  Handler function to create external storage storage vendor CR or resources. */
+  createPayload: CreatePayload<{}>;
+  /**  Handler function to validate the storage class page in order to move to the next step of wizard */
+  canGoToNextStep: CanGoToNextStep<{}>;
+  /** A function returing a promise, resolving any resources which should be created (if any) before creation of external sub-system */
+  waitToCreate?: (model: Model) => Promise<void>;
+};
+
+/** Props for `StorageClassWizardStepProps.component` to input the connection details of the external storage vendor. */
+export type StorageClassComponentProps<S extends ExternalState> = {
+  /** The state of the `StorageClassWizardStepProps.component` */
+  formState: S;
+  /** The callback for setting the state of `StorageClassWizardStepProps.component` */
+  setFormState: (field: keyof S, value: S[keyof S]) => void;
+  /** The object for registering components into `React Hook Form` */
+  control?: Control;
+};
+
+/**
+ *  @function CreatePayload<S>
+ *
+ *    @param {string} systemName
+ *    The name of the external storage system requried for the creation of the external custom resource.
+ *
+ *    @param {S extends ExternalState} state
+ *    The other fields of the create storage class form.
+ *
+ *    @param {ExtensionK8sModel} model
+ *    The model referring the `apiGroup`,`apiVersion` and `kind` of the external storage vendor's CRD.
+ *
+ *    @param {string} storageClassName
+ *    The name of the of the storage class.
+ *
+ *    @returns {Payload} An array of payloads of `Payload` type.
+ */
+export type CreatePayload<S extends ExternalState> = (payloadOptions: {
```

Comment:
```suggestion
export type CreatePayload<S extends ExternalState = ExternalState> = (payloadOptions: {
```
Although, I don't see the need for extends or to use type generics here.